pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1093,24 +1080,32 @@ protected IdealState validateLLCSegments(final TableConfig tableConfig, IdealSta
                     + "Old segment metadata {} has status DONE, but segments are still in CONSUMING state in ideal STATE",
                 tableNameWithType, partition, segmentId);
 
-            final int newSeqNum = segmentName.getSequenceNumber() + 1;
-            LLCSegmentName newLLCSegmentName =
-                new LLCSegmentName(segmentName.getTableName(), partition, newSeqNum, now);
-
+            LLCSegmentName newLLCSegmentName = makeNextLLCSegmentName(segmentName, partition, now);
             LOGGER.info("{}: Creating new segment metadata for {}", tableNameWithType,
                 newLLCSegmentName.getSegmentName());
 
             CommittingSegmentDescriptor committingSegmentDescriptor =
-                new CommittingSegmentDescriptor(segmentId, metadata.getEndOffset(), 0);
+                new CommittingSegmentDescriptor(segmentId, latestMetadata.getEndOffset(), 0);
             createNewSegmentMetadataZNRecord(tableConfig, segmentName, newLLCSegmentName, partitionAssignment,
                 committingSegmentDescriptor);
 
             onlineSegments.add(segmentId);
             consumingSegments.add(newLLCSegmentName.getSegmentName());
           }
           // else, the metadata should be IN_PROGRESS, which is the right state for a consuming segment.
-        } else if (isAllInstancesOffline(instanceStateMap)) {
-          // An in-progress segment marked itself offline for some reason the server could not consume.
+        } else { // no replica in CONSUMING state
+
+          // Possible scenarios: for any of these scenarios, we need to create new metadata IN_PROGRESS and new CONSUMING segment
+          // 1. all replicas OFFLINE and metadata IN_PROGRESS/DONE - a segment marked itself OFFLINE during consumption for some reason
+          // 2. all replicas ONLINE or mixture of ONLINE/OFFLINE and metadata DONE
+          // This could happen due to a race condition with the retention manager where we lost the new segment
+          // The retention manager saw the new segment metadata before it was updated in ideal state, and hence marked it for deletion
+          // later the new segment was updated in the ideal state, but the retention manager would clean it out from the metadata and ideal state, having marked it earlier
+          // So, in commmitSegmentMetadata:
```

Comment:
Is it better to do a check like this before applying the correction?

if (isallInstancesOffline() || !isTooSoonToCorrect(latestMetadata)) {
  // apply correction
}

this will protect us against other possible races that we don't know about yet. I can't think of any right now, but that is not to say that it does not exist.